crates/sui-snapshot/src/uploader.rs

```diff
@@ -0,0 +1,210 @@
+        local_staging_path: &std::path::Path,
+        remote_snapshot_config: ObjectStoreConfig,
+        interval_s: u64,
+        registry: &Registry,
+    ) -> Result<Self> {
+        let input_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::File),
+            directory: Some(input_path.to_path_buf()),
+            ..Default::default()
+        };
+        let local_staging_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::File),
+            directory: Some(local_staging_path.to_path_buf()),
+            ..Default::default()
+        };
+        Ok(StateSnapshotUploader {
+            local_input_path: input_path.to_path_buf(),
+            local_input_store: input_store_config.make()?,
+            local_staging_path: local_staging_path.to_path_buf(),
+            local_staging_store: local_staging_config.make()?,
+            remote_snapshot_store: remote_snapshot_config.make()?,
+            interval: Duration::from_secs(interval_s),
+            metrics: StateSnapshotUploaderMetrics::new(registry),
+        })
+    }
+
+    pub fn start(self) -> Sender<()> {
+        let (sender, mut recv) = channel::<()>();
+        let mut interval = tokio::time::interval(self.interval);
+        tokio::task::spawn(async move {
+            info!("State snapshot uploader loop started");
+            loop {
+                tokio::select! {
+                    _now = interval.tick() => {
+                        if let Ok(epochs) = find_missing_epochs_dirs(&self.remote_snapshot_store, SUCCESS_MARKER).await {
+                            self.metrics.first_missing_state_snapshot_epoch.set(epochs.first().cloned().map(|x| x as i64).unwrap_or(0));
+                            if let Err(err) = self.upload_state_snapshot_to_object_store(epochs).await {
+                                error!("Failed to upload state snapshot to remote store with err: {:?}", err);
+                            }
+                        } else {
+                            error!("Failed to find missing state snapshot in remote store");
+                        }
+                    },
+                    _ = &mut recv => break,
+                }
+            }
+        });
+        sender
+    }
+
+    async fn upload_state_snapshot_to_object_store(&self, missing_epochs: Vec<u32>) -> Result<()> {
+        let last_missing_epoch = missing_epochs.last().cloned().unwrap_or(0);
+        let local_checkpoints_by_epoch =
+            find_all_dirs_with_epoch_prefix(&self.local_input_store).await?;
+        let mut dirs: Vec<_> = local_checkpoints_by_epoch.iter().collect();
+        dirs.sort_by_key(|(epoch_num, _path)| *epoch_num);
+        for (epoch, db_path) in dirs {
+            if missing_epochs.contains(epoch) || *epoch >= last_missing_epoch {
+                let include_wrapped_tombstone =
+                    !ProtocolConfig::get_for_max_version().simplified_unwrap_then_delete();
```

Comment:
Please never use `get_for_max_version` unless you know this is specifically what you need. This will yield incorrect results as:

1. It reflects the locally max supported version of the running binary, not the version that the chain is on (the node can be upgraded before the network)
2. It will disregard chain specific config, as it sets a chain ID of `Chain::Unknown`, meaning in some cases you will get incorrect feature flag values. Notably, root state hash commitment is one such flag.

Instead, get the protocol config from `epoch_store.protocol_config()` or reconstruct by calling `ProtocolConfig::get_for_version(version: ProtocolVersion, chain: Chain)` and getting the version from the epoch start config and the chain from the authority state via `state().get_chain_identifier().chain()`

We should probably make this clearer, and yes there are several callsites that are doing this (narwhal for example). They need to be cleaned up